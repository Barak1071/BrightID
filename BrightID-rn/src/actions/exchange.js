import nacl from 'tweetnacl';
import { strToUint8Array, b64ToUint8Array } from './encoding';

import { setPublicKey2 } from './index';

export const GENERATE_MESSAGE = 'GENERATE_MESSAGE';

export const generateMessage = (publicKey2: Object) => (
  dispatch: Function,
  getState: Function,
) => {
  // set publickey from nearby connection
  publicKey2 = new Uint8Array(Object.values(publicKey2));
  dispatch(setPublicKey2(publicKey2));
  // generate timestamp
  const timestamp = Date.now();
  // obtain local public / secret keys
  const { publicKey, secretKey } = getState().main;
  // message (publicKey1 + publicKey2 + timestamp) signed by the private key of the user represented by publicKey1
  const message = strToUint8Array(
    publicKey.toString() + publicKey2.toString() + timestamp,
  );
  const genKeys = nacl.sign.keyPair();
  // console.warn(publicKey.toString());
  console.warn(new Buffer(message).toString());
  const sig1 = nacl.sign.detached(message, secretKey);
  const sig2 = nacl.sign.detached(message, genKeys.secretKey);
  // console.warn(publicKey instanceof Uint8Array);
  console.warn(nacl.sign.detached.verify(message, sig1, publicKey));
  console.warn(nacl.sign.detached.verify(message, sig2, genKeys.publicKey));

  // console.warn(signedMessage);
};
