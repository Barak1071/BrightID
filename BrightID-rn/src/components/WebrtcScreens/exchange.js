// @flow

import nacl from 'tweetnacl';
import { strToUint8Array } from '../../utils/encoding';

export const generateMessage = async () => (
  dispatch: Function,
  getState: Function,
) => {
  // set publickey from nearby connection
  // const {publicKey2} = new Uint8Array(Object.values(pk2));

  // dispatch(setPublicKey2(publicKey2));

  // generate timestamp
  const timestamp = Date.now();
  // obtain local public / secret keys
  const { publicKey, secretKey, publicKey2 } = getState().main;
  // message (publicKey1 + publicKey2 + timestamp) signed by the private key of the user represented by publicKey1
  if (!publicKey2) {
    return { error: 'no public key 2', msg: 'error - no public key 2' };
  }

  // this generates the message
  const message = strToUint8Array(
    publicKey.toString() + publicKey2.toString() + timestamp,
  );

  // console.warn(publicKey.toString());
  console.warn(Buffer.from(message).toString());
  const sig1 = nacl.sign.detached(message, secretKey);
  const verify = nacl.sign.detached.verify(message, sig1, publicKey);
  // console.warn(publicKey instanceof Uint8Array);
  console.warn(`message status: ${verify}`);
};

export const random = 'random';
