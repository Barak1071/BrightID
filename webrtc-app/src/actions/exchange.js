// @flow

import nacl from 'tweetnacl';
import { Buffer } from 'buffer';
import { strToUint8Array } from '../utils/encoding';

import { setPublicKey2, setPairingMessage } from './index';

export const GENERATE_MESSAGE = 'GENERATE_MESSAGE';

export const genMsg = () => (dispatch: Function, getState: Function) => {
  // set publickey from nearby connection
  const { publicKey, secretKey, publicKey2 } = getState();

  // generate timestamp
  const timestamp = Date.now();
  // obtain local public / secret keys
  // message (publicKey1 + publicKey2 + timestamp) signed by the private key of the user represented by publicKey1
  const message = strToUint8Array(
    publicKey.toString() + publicKey2.toString() + timestamp,
  );

  dispatch(
    setPairingMessage({
      msg: message,
      msgStr: Buffer.from(message).toString(),
    }),
  );

  // testing signed message
  const genKeys = nacl.sign.keyPair();
  // console.warn(publicKey.toString());
  console.log(Buffer.from(message).toString());
  const sig1 = nacl.sign.detached(message, secretKey);
  const sig2 = nacl.sign.detached(message, genKeys.secretKey);
  // console.log(publicKey instanceof Uint8Array);
  console.log(nacl.sign.detached.verify(message, sig1, publicKey));
  console.log(nacl.sign.detached.verify(message, sig2, genKeys.publicKey));

  // console.warn(signedMessage);
};
