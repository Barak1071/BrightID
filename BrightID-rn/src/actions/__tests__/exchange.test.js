import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import nacl from 'tweetnacl';

import { generateMessage } from '../exchange';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe('async action', () => {
  test('sets up dummy data without userData', () => {
    const keyPair1 = nacl.sign.keyPair();
    let keyPair2 = {
      publicKey: new Uint8Array([
        187,
        140,
        200,
        233,
        104,
        179,
        188,
        241,
        76,
        240,
        229,
        54,
        242,
        103,
        201,
        4,
        131,
        44,
        118,
        37,
        204,
        6,
        161,
        35,
        191,
        33,
        47,
        175,
        32,
        183,
        174,
        40,
      ]),
    };
    // mocking async storage conversion
    keyPair2 = JSON.stringify(keyPair2);
    keyPair2 = JSON.parse(keyPair2);

    const store = mockStore({ main: keyPair1 });

    store.dispatch(generateMessage(keyPair2.publicKey));
    expect(store.getActions()).toMatchSnapshot();
  });
});
