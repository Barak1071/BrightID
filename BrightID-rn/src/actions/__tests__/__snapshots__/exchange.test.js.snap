// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`async action sets up dummy data without userData 1`] = `
Array [
  Object {
    "publicKey2": Uint8Array [
      187,
      140,
      200,
      233,
      104,
      179,
      188,
      241,
      76,
      240,
      229,
      54,
      242,
      103,
      201,
      4,
      131,
      44,
      118,
      37,
      204,
      6,
      161,
      35,
      191,
      33,
      47,
      175,
      32,
      183,
      174,
      40,
    ],
    "type": "PUBLICKEY2",
  },
]
`;
